### PR TITLE
Make the persistence-service always available, regardless of solution size.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is a more aggressive solution htan https://github.com/dotnet/roslyn/pull/15042.

In this branch we remove everything related to our heuristic around determining SolutionSize.  This heuristic causes several issues around higher level services which use the persistence service.  Namely, it presumes that it's ok to just not store data if a solution is 'small enough' where 'small enough' is defined as <50MB of source code.  

However, this approach fails to address that that a solution might have less code than that, but may still use the persistence service to cache information that is extremely costly to produce (both in CPU and memory).

For example, our metadata indices are quite costly, and should not be recomputed every time a solution is opened, just because it only contains 49MB of code.  Similarly, if we were to store things like nuget caches in this index, it would be a very bad idea for us to have to keep on redownloading things for small projects.  

I'm not sure at all what benefit this heuristic serves.  The persistence service is a publicly available Roslyn service that does not document this behavior at all.  Services higher up in the stack should be able to trust that this service works effectively regardless of solution size.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @srivatsn @heejaechang @panopticoncentral

**Comment by heejaechang** [comment]
@CyrusNajmabadi does this affect both VS and OOP? esent uses about 50MB itself. turned down turning it up by default from upper management since we can't get back 50MB VM increase RPS regression from somewhere else.

if you can get the exception, I am fine. otherwise, you need to figure out how to bring 50MB back before you can check this in.

**Comment by CyrusNajmabadi** [comment]
> turned down turning it up by default from upper management since we can't get back 50MB VM increase RPS regression from somewhere else.

I'm confused by this.  Don't you get the 50MB VM hit no matter what if your solution size is large enough?

Are we saying that we literally just hardcoded in a value to make RPS tests pass, even though in practice people will have this overhead?  

If so, that seems very suspicious to me.  It makes it so our numbers are green, but really that's only the case because we made it so the test only measures a very specific scenario that is not really reflective of how people will experience Roslyn on their machines.

If this 50MB of VM is a problem, then it's a problem in all cases, not just the specific cases that RPS tests.  Hardcoding in values to make RPS pass, while still allowing us to use this doesn't seem ok to me.  @srivatsn can you weigh in here?  Are we ok with making changes just for RPS purposes, even if the end experience won't incur that cost?

If so, can we just add a switch for stating if we're running in RPS so that we don't use esent at all?  Then our numbers will stay good for RPS, while allowing us to have a reliable persistence service that our features can depend on so they don't have to recompute data every time a solution is opened.

**Comment by CyrusNajmabadi** [comment]
@KevinH-MS Is there a way to know if you're running in RPS?  We'd like to put in some code to disable something that is causing memory overhead so we don't need an exception.  Right now we have a weak heuristic that tries to determine if we're in RPS, but it has issues, causing problems later on.  I'd rather get rid of the heuristic and make the check exact.  That will allow us to keep RPS green, while keeping correctness for our features.

**Comment by heejaechang** [comment]
solution size is determined not by RPS test, but calculating size of data cached in memory if no persistent service is there vs overhead of esent + data left in memory.

and in real life, it does make things like console or small solution, not hit by esent overhead.

**Comment by CyrusNajmabadi** [comment]
> and in real life, it does make things like console or small solution, not hit by esent overhead.

Except now you're hit by all the costs of the features that expected the persistence service to work, but which now need to recompute everything.

This is exceptionally strange.  If 50MB is an unacceptable memory hit (which RPS seems to indicate), then the esent solution is a problem *regardless of this heuristic*.  All the heuristic does is make RPS pass by pretending the 50MB VM hit isn't there.  But on large solutions *where memory pressure is a concern* you now face this hit.

Making it so that small solutions stay small isn't actually helpful.  They're already small.  They have tons of memory available to them.  The concern is about large projects, and in that regard we will consume this memory.  We just don't report that during testing, so it appears as if everything is ok.

**Comment by heejaechang** [comment]
in our code, we code that users persistent service assumes it can fail, and if it fail, it deal with it. some cache them in memory, so just let it go and re-calculate.

and when all those are just cached in memory. we found line where that size is about same as esent overhead. and only when solution size is at that level, we switch on esent.

so those memory spent in cache now goes down and esent overhead takes up that memory and after that memory increase stop.

basically, up until the solution size line, as solution increase, memory used by cache goes up, once size hit to solution size line, memory stop grow and maintain that line.

**Comment by heejaechang** [comment]
by the way, persistent service is host service, it is public since someone who want to implement new workspace host can provide their own implementation that workspace can use. it is not workspace service such as find all reference service which third party supposed to get from us and use.

**Comment by CyrusNajmabadi** [comment]
> in our code, we code that users persistent service assumes it can fail, 

No.  We code with the presumption that failure is rare, not the 100% case for some solutions.   We understand that persistence may just fail for something like IO.  Fine, that's a rare case, so it should be treated rarely.  We also get that data may not be read back.  Either because of IO again, or because the versions have changed on disk.  Again, both rare cases.  

The expectation was that for the normal use case the service would work.  As we expect for the normal use case or all Roslyn APIs.  For example, symbol reading APIs don't fail if a solution is below a certain size.  Roslyn Features don't fail if you have too few projects.  

This is a case of a service failing 100% of the time for some large set of solutions.  This is not normal, and *as you have claimed yourself* appears to have only been done to make it so we can pass RPS by not measuring the 50MB memory hit.   That decision, which affects a core service that many layers use, was not communicated.

Imagine if we did the same thing for other roslyn features.  Imagine if FAR simply said "your solution is too small, so you get no results back" thus ensuring that our FAR numbers in RPS were always good.  We could hten claim that that was an implementation detail.  But it would be devastating for the layers that exist on top of FAR and depend on it working.

Again, we designed features with the expectation that the lower layers would work normally the majority of the time, and certainly for normal solutions.  The idea that this is not the case makes all of our higher designs that depend on this behavior *suspect*.  That is *extremely* concerning.  As concerning as if you were to learn *just now* about the OOP guys changing core behavior without properly communicating that they were doing so, makign all your decisions that you layered on top of their original designs suspect.

**Comment by heejaechang** [comment]
I am not sure what rarely affect what we need to do. when it happens either we need to handle it or not. if we handle it, it doesn't matter how often it happens, it handle it. if it doesn't it doesn't. when it doesn't, we still need to make sure we work as expected since persistent service is host service. persistent service is something we expect third party provide to us. we need to guard us from them.

**Comment by CyrusNajmabadi** [comment]
>  it is not workspace service such as find all reference service which third party supposed to get from us and use.

```c#
    /// <summary>
    /// This service allows you to persist information relative to solution, projects and documents.
    /// </summary>
    public interface IPersistentStorageService : IWorkspaceService
    {
        IPersistentStorage GetStorage(Solution solution);
    }
```

1. it is a workspace service.  
2. there was absolutely no documentation of communication that this service was not intended for anyone to use.  

Had IDE been aware of these non-communicated issues, then either:

1.  IDE would *not* have used this persistence service for our feature work.  
2. IDE would have insisted on providing a way for this persistence service to function as it was originally communicated to do.

**Comment by CyrusNajmabadi** [comment]
> I am not sure what rarely affect what we need to do. when it happens either we need to handle it or not. if we handle it, it doesn't matter how often it happens, it handle it.

This is absolutely, 100% not the case.

It absolutely matters how often it happens.  And it matters *deeply* if it's going to happen 100% of the time for some class of solutions.  I'm extremely concerned that you cannot see that, and you cannot see how it would be important to document or communicate these issues.  

If you cannot see the difference between a service that succeeds nearly 100% of the time, and which might fail rarely, versus a service that fails 100% of the time for some common cases, that's a serious problem.  

We design our code to work properly even in the rare case, but to work optimally in the common case.  Thus, it is paramount for people to know what the common case is.   If you've changed what the common case is, then code may still work properly, but may not work optimally at all.  That's a very big deal.
